### ISSUE TITLE
[ruby] grpc protoc plugin could generate broken request/response namespaces

### ISSUE THREAD

**Issue by dazuma** [grpc/grpc#23746]
### What version of gRPC and what language are you using?

gRPC 1.30.2 on Ruby on Ubuntu bionic. Specifically the protoc plugin distributed as part of https://rubygems.org/gems/grpc-tools/versions/1.30.2

### What did you do?

Running the grpc protoc plugin on the [grafeas.v1 proto](https://github.com/googleapis/googleapis/tree/master/grafeas/v1) (as part of client library codegen).

The result looks like [this](https://github.com/googleapis/google-cloud-ruby/blob/0ada439dee22d319baf50c0e82174a4b8ce212db/grafeas-v1/lib/grafeas/v1/grafeas_services_pb.rb)

An excerpt with comments removed:

```ruby
require 'grpc'
require 'grafeas/v1/grafeas_pb'

module Grafeas
  module V1
    module Grafeas
      class Service

        include GRPC::GenericService

        self.marshal_class_method = :encode
        self.unmarshal_class_method = :decode
        self.service_name = 'grafeas.v1.Grafeas'

        rpc :GetOccurrence, Grafeas::V1::GetOccurrenceRequest, Grafeas::V1::Occurrence  # <-- error here

# and additional lines...
```

The indicated line results in an error:
```
/path/to/grafeas_services_pb.rb:48:in `<class:Service>': uninitialized constant Grafeas::V1::Grafeas::V1 (NameError)
Did you mean?  Grafeas::V1
```

This is happening because when Ruby does constant lookup on `Grafeas::V1::GetOccurrenceRequest`, it searches lexically up the namespace chain for `Grafeas` and instead of the toplevel `Grafeas` module, the first one it finds is actually the `Grafeas::V1::Grafeas` module.

The fix is to prefix the namespace with `::` which tells Ruby to search from the toplevel instead. That is, the above line should look like:

```ruby
        rpc :GetOccurrence, ::Grafeas::V1::GetOccurrenceRequest, ::Grafeas::V1::Occurrence
```

This is always safe to do whenever you are providing a "fully qualified" name, as you are above. (In fact, for generated code, it is good practice _always_ to do it, specifically to avoid this bug.)

Note that this is probably a regression from #23501 which, it seems, changed the behavior so that fully-qualified request and response class names are always generated.

grpc 1.30.1 generates:

```ruby
# Works because GetOccurrenceRequest is a "relative" class name
# i.e. it resolves correctly to Grafeas::V1::GetOccurrenceRequest
# and Grafeas::V1::Occurrence.
# (However, this version suffered from #23490 which was tangetially related)
rpc :GetOccurrence, GetOccurrenceRequest, Occurrence
```

grpc 1.30.2 generates:

```ruby
# Does not work. Resolves incorrectly to 
# Grafeas::V1::Grafeas::V1::GetOccurrenceRequest and
# Grafeas::V1::Grafeas::V1::Occurrence
rpc :GetOccurrence, Grafeas::V1::GetOccurrenceRequest, Grafeas::V1::Occurrence
```

What we want:

```ruby
# Explicitly do absolute rather than relative lookup. Resolves correctly to 
# Grafeas::V1::GetOccurrenceRequest and Grafeas::V1::Occurrence
rpc :GetOccurrence, ::Grafeas::V1::GetOccurrenceRequest, ::Grafeas::V1::Occurrence
```

/cc @stanley-cheung

**Comment by stanley-cheung** [comment]
Probably dupe of #23672?

**Comment by dazuma** [comment]
@stanley-cheung Ah yes, that does look like the same issue. Consider this a remedy suggestion.

**Comment by stanley-cheung** [comment]
Thanks for confirming. We are working on this.

**Comment by stanley-cheung** [comment]
@dazuma A question: does it make more sense to go back to the 1.30.1 behavior to generate
```
rpc :GetOccurrence, GetOccurrenceRequest, Occurrence
```
if the classes are all in the same package? Instead of
```
rpc :GetOccurrence, ::Grafeas::V1::GetOccurrenceRequest, ::Grafeas::V1::Occurrence
```

**Comment by splittingred** [comment]
@stanley-cheung In my opinion, it would be preferable to always use the absolute name pathing, so that we always ensure resolution is correct. In the former, this can cause issues with cross-package names.

**Comment by dazuma** [comment]
@stanley-cheung If we can guarantee that the classes are in the same namespace, then it _is safe_ to use the relative names because the constant lookup algorithm looks there first. However, for generated code, I say it's still _preferred_ to use the full path, if for nothing else then to stave off future bugs if a class ever gets moved to a different namespace.

**Comment by stanley-cheung** [comment]
@dazuma The fix should be in `1.31.1` now. Can you please check it out? Thanks!
